src/declarativedbusinterface.cpp

```diff
@@ -413,6 +413,103 @@ void DeclarativeDBusInterface::componentComplete()
 }
 
+static QVariant unwind(const QVariant &val)
+{
+    QVariant res;
+
+    int type = val.userType();
+
+    if( type == val.type() ) {
+        /* Already is built-in qt type, use as is */
+        res = val;
+    }
+    else if( type == qMetaTypeId<QDBusVariant>() ) {
+        /* Convert QDBusVariant to QVariant */
+        res = unwind(val.value<QDBusVariant>().variant());
+    }
+    else if( type == qMetaTypeId<QDBusObjectPath>() ) {
+        /* Convert QDBusObjectPath to QString */
+        res = val.value<QDBusObjectPath>().path();
+    }
+    else if( type == qMetaTypeId<QDBusSignature>() ) {
+        /* Convert QDBusSignature to QString */
+        res =  val.value<QDBusSignature>().signature();
+    }
+    else if( type == qMetaTypeId<QDBusArgument>() ) {
+        /* Try to deal with everything QDBusArgument could be ... */
+
+        const QDBusArgument & arg = val.value<QDBusArgument>();
+
+        switch( arg.currentType() ) {
+        case QDBusArgument::BasicType:
+            /* Most of the basic types should be convertible to QVariant.
+             * Recurse anyway to deal with object paths and the like. */
+            res = unwind(arg.asVariant());
+            break;
+
+        case QDBusArgument::VariantType:
+            /* Try to convert to QVariant. Recurse to check content */
+            res = unwind(arg.asVariant().value<QDBusVariant>().variant());
+            break;
+
+        case QDBusArgument::ArrayType:
+            /* Convert dbus array to QVariantList */
+            {
+                QVariantList list;
+                arg.beginArray();
+                while( !arg.atEnd() ) {
+                    QVariant tmp = arg.asVariant();
+                    list.append(unwind(tmp));
+                }
+                arg.endArray();
+                res = list;
+            }
+            break;
+
+        case QDBusArgument::StructureType:
+            /* Convert dbus struct to QVariantList */
+            {
+                QVariantList list;
+                arg.beginStructure();
+                while( !arg.atEnd() ) {
+                    QVariant tmp = arg.asVariant();
+                    list.append(unwind(tmp));
+                }
+                arg.endStructure();
+                res = QVariant::fromValue(list);
+            }
+            break;
+
+        case QDBusArgument::MapType:
+            /* Convert dbus dict to QVariantMap */
+            {
+                QVariantMap map;
+                arg.beginMap();
+                while (!arg.atEnd()) {
+                    arg.beginMapEntry();
+                    QVariant key = arg.asVariant();
+                    QVariant val = arg.asVariant();
+                    map.insert(unwind(key).toString(), unwind(val));
+                    arg.endMapEntry();
+                }
+                arg.endMap();
+                res = map;
+            }
+            break;
+
+        default:
+            /* Unhandled types produce invalid QVariant */
```

Comment:
From dbus spec:"A DICT_ENTRY works exactly like a struct, but rather than parentheses it uses curly braces, and it has more restrictions. The restrictions are: it occurs only as an array element type ..."

In qt arrays that contain dict entries -> MapType, i.e. MapEntryType should occur only within MapType context - where it is now handled via beginMapEntry() ... endMapentry().